pkg/descheduler/strategies/toomanyrestarts_test.go

```diff
@@ -0,0 +1,207 @@
+				{
+					RestartCount: 5 * i,
+				},
+			},
+			ContainerStatuses: []v1.ContainerStatus{
+				{
+					RestartCount: 10 * i,
+				},
+				{
+					RestartCount: 10 * i,
+				},
+			},
+		}
+		pods = append(pods, *pod)
+		i++
+	}
+
+	// The following 4 pods won't get evicted.
+	// A daemonset.
+	pods[6].ObjectMeta.OwnerReferences = test.GetDaemonSetOwnerRefList()
+	// A pod with local storage.
+	pods[7].ObjectMeta.OwnerReferences = test.GetNormalPodOwnerRefList()
+	pods[7].Spec.Volumes = []v1.Volume{
+		{
+			Name: "sample",
+			VolumeSource: v1.VolumeSource{
+				HostPath: &v1.HostPathVolumeSource{Path: "somePath"},
+				EmptyDir: &v1.EmptyDirVolumeSource{
+					SizeLimit: resource.NewQuantity(int64(10), resource.BinarySI)},
+			},
+		},
+	}
+	// A Mirror Pod.
+	pods[8].Annotations = test.GetMirrorPodAnnotation()
+
+	return pods
+}
+
+func TestRemovePodsHavingTooManyRestarts(t *testing.T) {
+
+	createStrategy := func(enabled, includingInitContainers bool, restartThresholds int32) api.DeschedulerStrategy {
+		return api.DeschedulerStrategy{
+			Enabled: enabled,
+			Params: api.StrategyParameters{
+				PodsHavingTooManyRestarts: api.PodsHavingTooManyRestarts{
+					PodeRestartThresholds:   restartThresholds,
+					IncludingInitContainers: includingInitContainers,
+				},
+			},
+		}
+	}
+
+	node := test.BuildTestNode("node1", 2000, 3000, 10)
+	pods := initPods(node)
+
+	fakeClient := &fake.Clientset{}
+
+	fakeClient.Fake.AddReactor("list", "pods", func(action core.Action) (bool, runtime.Object, error) {
+		return true, &v1.PodList{Items: pods}, nil
+	})
+
+	tests := []struct {
+		description             string
+		pods                    []v1.Pod
+		strategy                api.DeschedulerStrategy
+		expectedEvictedPodCount int
+		maxPodsToEvict          int
+	}{
+		{
+			description: "Not a PodsHavingTooManyRestarts strategy, no pod evictions",
+			strategy: api.DeschedulerStrategy{
+				Enabled: true,
+				Params: api.StrategyParameters{
+					NodeAffinityType: []string{
+						"requiredDuringSchedulingRequiredDuringExecution",
+					},
+				},
+			},
+			expectedEvictedPodCount: 0,
+			maxPodsToEvict:          0,
+		},
+		{
+			description:             "Strategy is disabled, no pod evictions",
+			strategy:                createStrategy(false, true, 10),
+			expectedEvictedPodCount: 0,
+			maxPodsToEvict:          0,
+		},
+		{
+			description:             "All pods have total restarts under threshold, no pod evictions",
+			strategy:                createStrategy(true, true, 10000),
+			expectedEvictedPodCount: 0,
+			maxPodsToEvict:          0,
+		},
+		{
+			description:             "One pod have total restarts bigger than threshold, 6 pod evictions",
+			strategy:                createStrategy(true, true, 1),
+			expectedEvictedPodCount: 6,
+			maxPodsToEvict:          0,
+		},
+		{
+			description:             "Nine pods have total restarts equals threshold(includingInitContainers=true), 5 pods evictions",
```

Comment:
The strategy has the following condition for deciding if the threshold is met:
```
if strategy.Params.PodsHavingTooManyRestarts.IncludingInitContainers {
	if restarts+initRestarts <= strategy.Params.PodsHavingTooManyRestarts.PodeRestartThresholds {
		continue
	}
} else if restarts <= strategy.Params.PodsHavingTooManyRestarts.PodeRestartThresholds {
	continue
}
```
Which means a pod is evicted only when the number of restarts is bigger than the threshold. Is this the intention? I would assume a pod to be evicted as soon as the threshold is met.